html/webappapis/dynamic-markup-insertion/opening-the-input-stream/tasks.window.js

```diff
@@ -0,0 +1,114 @@
+    // event, which is necessary to make sure the frame's document doesn't have
+    // a parser associated with it.
+    frame.src = "/common/blank.html";
+    t.add_cleanup(() => frame.remove());
+    frame.onload = t.step_func(() => {
+      // Make sure there is no parser. Firefox seems to have an additional
+      // non-spec-compliant readiness state "uninitialized", so test for the
+      // two known valid readiness states instead.
+      assert_in_array(frame.contentDocument.readyState, ["interactive", "complete"]);
+      testBody(t, frame, doc => doc.open());
+    });
+  }, `document.open() and tasks (${description})`);
+}
+
+taskTest("timeout", (t, frame, open) => {
+  let happened = false;
+  // Work around the lint script, since we can't use step_timeout here.
+  frame.contentWindow['setTimeout'](() => happened = true, 100);
+  open(frame.contentDocument);
+  t.step_timeout(() => {
+    assert_true(happened);
+    t.done();
+  }, 200);
+});
+
+taskTest("window message", (t, frame, open) => {
+  let counter = 0;
+  frame.contentWindow.postMessage(undefined, "*");
+  open(frame.contentDocument);
+  frame.contentWindow.postMessage(undefined, "*");
+  frame.contentWindow.onmessage = t.step_func(e => {
+    assert_equals(e.data, undefined);
+    counter++;
+    assert_less_than_equal(counter, 2);
+    if (counter == 2) {
+      t.done();
+    }
+  });
+});
+
+taskTest("canvas.toBlob()", (t, frame, open) => {
+  const canvas = frame.contentDocument.body.appendChild(frame.contentDocument.createElement("canvas"));
+  canvas.toBlob(t.step_func_done());
+  open(frame.contentDocument);
+});
+
+taskTest("MessagePort", (t, frame, open) => {
+  frame.contentWindow.eval(`({ port1, port2 } = new MessageChannel());`);
+  frame.contentWindow.port2.onmessage = t.step_func_done(ev => {
+    assert_equals(ev.data, "Hello world");
+  });
+  frame.contentWindow.port1.postMessage("Hello world");
+  open(frame.contentDocument);
+});
+
+taskTest("Promise rejection", (t, frame, open) => {
+  // There is currently some ambiguity on which Window object the
+  // unhandledrejection event should be fired on. Here, let's account for that
+  // ambiguity and allow event fired on _any_ global to pass this test.
+  // See:
+  // - https://github.com/whatwg/html/issues/958,
+  // - https://bugs.webkit.org/show_bug.cgi?id=187822
+  const promise = frame.contentWindow.eval("Promise.reject(42);");
+  open(frame.contentDocument);
+  const listener = t.step_func_done(ev => {
+    assert_equals(ev.promise, promise);
+    assert_equals(ev.reason, 42);
+  });
+  frame.contentWindow.onunhandledrejection = listener;
+  window.onunhandledrejection = listener;
+});
+
+taskTest("marquee start", (t, frame, open) => {
+  const doc = frame.contentDocument;
+  const marquee = doc.body.appendChild(doc.createElement("marquee"));
+  open(frame.contentDocument);
+  marquee.addEventListener("start", t.step_func_done());
```

Comment:
I like this one. If you want one which is a bit special in practice try a new audio element and changing the volume attribute to 0.5. That should queue a task to fire volumechange. But implementations do something special to emulate that per-instance task source that media elements are supposed to have, and I wouldn't be surprised if it's different from other things tested here.